### PR TITLE
Add entry for Kotlin

### PR DESCRIPTION
**Pull request by IvoMiller**
- [x] Are you happy for the content of this change to be publicly visible?
- [x] Are all new or updated entries marked as `isNew` = `true`?
- [x] Does `radar.csv` render correctly when viewed on Github?
- [x] Does the updated [tech radar](https://radar.thoughtworks.com/?documentId=https%3A%2F%2Fraw.githubusercontent.com%2Fred-gate%2FTech-Radar%2Fkotlin%2Fradar.csv) display as you expect?

This is a proposal for raising Kotlin for consideration when writing code which needs to compile to Java Byte Code.
We have already run an experiment within Team Pegasus which makes us feel like this is a very viable option, and we list a bunch of considerations below.
While there is a certain amount of justification around why Kotlin might make sense for flyway plugins in particular, that particular use case is not what is under discussion here. Rather this is about whether Kotlin is a viable option at Redgate in general.

## Context
Team Pegasus are looking to rewrite the `flyway-dev` command lines as plugins for Flyway.
The top layer of Flyway plugins needs to be in Java or a language which compiles down to Java byte code (e.g. Kotlin).
There isn't all that much code at the top `flyway-dev` CLI layer, and making Java interface with our existing CLIs feels roundabout and undesirable.
We therefore think that it would be best to rewrite our existing CLI code in Java/Kotlin instead of C#.
This will take some time but can be done incrementally.

The main impact is the need for Team Pegasus to upskill in Java/Kotlin.
We have not yet reached a conclusion at the group level around language choice between Java and Kotlin. We have had some constrctive discussions between teams but we want to raise this for wider discussion to see whether Kotlin is an option we are allowed to consider.

Here follows a summary of why we on Team Pegasus believe Kotlin is a language worth considering. We have mixed in some of the justification for use in flyway with the wider justification around the consideration of the language, because the context feels important.

## TL; DR;
We believe we will write safer code and release fewer bugs if we chose Kotlin over Java

We expect there to be some pushback on the basis that this is a new language but we believe that this decision is like picking typescript over javascript. It provides a layer of safety over the underlying language which contains a bunch of pitfalls as a result of its long history, and importantly it does not have a distinct ecosystem. Java and Kotlin can exist side by side, within the same project/module/package if need be, and Kotlin can use all the Java libraries. While we have a couple of experienced Java developers on the team, we believe that adopting Kotlin is less of a learning curve for Team Pegasus as a whole than adopting Java.

## Cognitive load of a new language
Proposing a new language should never be undertaken lightly and there are a few areas where there is potential for cognitive load
* Abstraction - a language has some core set of abstractions around which it is based. If these are unfamiliar, it can take some time to get into the mindset. This isn't really an issue in the case of Kotlin as it is very much in the same family as Java and C#.
* Syntax - any language comes with its own syntax which needs to be learnt. Kotlin has a few peculiarities (e.g. currying), but it broadly looks like a combo of Java/C# and Typescript, so should look pretty familiar to members of Pegasus or Geese; syntax alone is not a very big barrier. The idioms are again similar to Java/C#.
* Ecosystem - by far the biggest potential source of cognitive load in a new language is the ecosystem: how builds work, how dependencies are handled, common libraries which are used to solve standard enterprise problems, ...  Kotlin does not have a distinct ecosystem and piggybacks on all the java libraries. We can use Maven for build and all the same libraries as the Flyway engine.

## Modern, concise, and safe
Kotlin markets itself as modern, concise, and safe.

In terms of modern, one nice feature of Kotlin is that you just pick the version of Java to compile down to, so it is always possible to use the latest version of the language, without being constrained by what customers have installed.
Concise is nice, and Kotlin does have a lot of useful language features missing in Java (e.g. extension methods), though maybe not as important as familiarity for other Redgaters. There are places where the abstractions over Java can avoid a lot of confusing boilerplate though (e.g. no need to use `final` keyword or defensive copying, no need to handle primitives and objects separately).
The real selling point of Kotlin from our perspective is safety though. A major aspect of this is strict nullability, which we are all very used to from C# and Typescript. We have seen instances of NullPointerException being thrown by the Flyway engine in the past, so this is something that we already fall foul of and would be worth defending against at compile time. Java also returns null in different places from C# in fairly standard situations (e.g. Map), and it is very easy for developers coming from C# to write code liable to produce null exceptions (this was at one time an issue in Redgate DLMA plugins for TeamCity/Jenkins/Bamboo). While there are some Java features which allow for localised improvements to null handling (Optional and @NotNull) there is nothing in the language to enable compile-time null-safety.
Kotlin also defends well against side effect code and unanticipated code extension, naturally enforcing a lot of Java good practices which are otherwise quite verbose to enforce.
Collections are immutable by default and there is no need for defensive copying.
Parameters cannot be reassigned and variables and fields are defined as `var` (re-assignable) or `val` (not re-assignable). You don't have to use `final` everywhere to defend against this. 
Classes are also final by default and not subject to unanticipated extension or access via reflection.

Here is an image from a census of Kotlin users taken by Jetbrains in 2020 which demonstrates how null safety is the most common reason for adoption of the language
image

## Other Java pitfalls that Kotlin protects against
No boxing/unboxing - in Java primitives cannot be handled using generics and have to be boxed into object equivalents. In certain scenarios this leads to a lot of boilerplate code, and there are some subtle associated gotchas. Kotlin handles all types as objects and everything is much simpler.
Associated with this, the array versus list issue is no longer a problem in Kotlin. Arrays in Kotlin are invariant.
While Kotlin still operates generics by erasure, it doesn't allow raw types, and doesn't use wildcards. It improves on generic types (reified type parameters), making them usable in code more like how you'd expect in C#.
Kotlin has proper function types.
Kotlin does not have checked exceptions (which have been a particular headache in Java since the introduction of lambdas in Java 8, albeit alleviated by lombok SneakyThrows)

## No obvious technical constraints
There are no obvious technical constraints for use within flyway. These plugins would not be open source. Using Kotlin adds two jars as dependencies, both of them very small.
The Kotlin license is compatible with the Flyway license.
IntelliJ tooling around Kotlin is excellent (hardly surprising as IntelliJ is written in Kotlin).

## Examples of other companies taking this approach
There is plenty of evidence of large well-respected companies choosing Kotlin over Java.
Google has backed Kotlin for all things Android and is a force behind Kotlin's adoption.
Meta have recently been converting all of their Java to Kotlin. [From zero to 10 million lines of Kotlin](https://engineering.fb.com/2022/10/24/android/android-java-kotlin-migration/)

## Experiment
We ran an experiment writing a Flyway plugin in Kotlin in Team Pegasus, fully coded, unit tested, and manually tested, using maven modules.
All of the service loading and reflection used in Flyway engine in the loading of the plugin worked fine.
All of the team were involved in either the coding or the reviewing of the code. The team was able to get up and running in the language pretty fast. The few gotchas raised in review were Java gotchas rather than specifically Kotlin.

## Impact on hiring/interviewing?
Were this a success, we don't feel that we need to be specifically hiring Kotlin devs.
Kotlin should be comparatively easy for Java and C# devs to pick up.
When we are hiring devs at Redgate, we are not testing them in every language we use.

## How to measure the success of the trial, should it go ahead?
We welcome input here.
Success feels like successfully shipping and maintaining some Kotlin code. If it were used for the flyway-dev flyway plugins, it is highly unlikely that the team writing these plugins will cease to be the owners within the next year or so, as there is a lot of associated domain knowledge.
Inevitably, if we did back out and go back to Java in the future, some code rewriting would be needed. While IntelliJ can auto-convert Java to Kotlin, it cannot do the reverse. That said, the way the languages work is very similar, so it would be broadly possible to rewrite like for like, with the same libraries and infrastructure.

### CODE REVIEW

**Comment by clivetong** [comment]
`to decompile to Java Byte Code` - I assume you mean `compile to Java Byte Code`

**Comment by clivetong** [comment]
Out of interest, would you say Kotlin is tighter to Java than say F# to C#. For F# you often have to write code specially to act as the bridge between the F# and C# worlds - different datatypes (like discriminated unions in F# but not in C#), different representations of functions. I think you were hinting at that by talking about Typescript and JavaScript, but it's just that TS is a superset of JS (syntactically I think) whereas that's not the case with Kotlin?

**Comment by IvoMiller** [comment]
> Out of interest, would you say Kotlin is tighter to Java than say F# to C#. For F# you often have to write code specially to act as the bridge between the F# and C# worlds - different datatypes (like discriminated unions in F# but not in C#), different representations of functions. I think you were hinting at that by talking about Typescript and JavaScript, but it's just that TS is a superset of JS (syntactically I think) whereas that's not the case with Kotlin?

So with the caveat that I have not really used F#:
* I believe that Kotlin is closer to Java than F# to C#. All the types map very obviously to each other. Interop works very smoothly both ways. In our experiment we found no need for any form of bridging code, and I can't see why we would ever generate any.
* You are right that it is not a strict superset of Java like typescript is of javascript, but I still believe that is a better parallel than C# vs F#, because Kotlin is fundamentally a very similar language to Java and the interop is so good

**Comment by Greg-Smulko** [comment]
There are lots of valuable comments from @DoodleBobBuffPants and @madninjaskillz in https://github.com/red-gate/FlywayDesktop/issues/10463 predating this issue by 2+ weeks and a good discussion there.

I must admit I found it surprising that it wasn't linked here.

**Comment by IvoMiller** [comment]
> There are lots of valuable comments from @DoodleBobBuffPants and @madninjaskillz in [red-gate/FlywayDesktop#10463](https://github.com/red-gate/FlywayDesktop/issues/10463) predating this issue by 2+ weeks and a good discussion there.
> 
> I must admit I found it surprising that it wasn't linked here.

Apologies. Maybe it should have been. As that was a draft and more specifically about whether to use it for the plugins (which is a separate albeit very related discussion) I didn't think to link to it.

**Comment by TheEadie** [comment]
We have decided not to proceed with trialing Kotlin for new JRE code. There are a few reasons we believe this is the right decision.

1. We don't expect any more new JRE code to be written beyond writing plugins for Flyway. New projects will choose C# as a backend language.
2. Flyway's open source project is currently in Java - we don't feel we should mix languages within the open source project.
3. We want the option to move plugins between the closed source repo and the open source repo without having to rewrite the code.
4. We don't have a desire or plan to migrate the current Flyway code from Java to Kotlin.
5. The teams working on Flyway will need to understand the idioms of Java to contribute to the main codebase. Therefore we don't want to increase the cognative load for these teams to understand another language's idioms as well.

We will reassess this if:

1. There is a clear sign that the wider Java community is moving away from Java and towards Kotlin. Specifically the move is big enough for us to consider moving the Flyway open source codebase to Kotlin.
